data race

Yo man, I cloned the repo and added this test:

```golang
func TestRace(t *testing.T) {
	ctx := context.Background()
	ab := NewAtomicCircularBuffer2(1000)
	event := createTestEvent("aaaaaaaaaaaaaaaaaaaaaaa", 1)

	go func() {
		for range 1000000 {
			ab.QueryEvents(ctx, nostr.Filter{IDs: []string{"aaa"}})
		}
	}()

	for range 1000000 {
		ab.SaveEvent(ctx, event)
	}
}
```

Then run `go test -race -run=^TestRace$`
This is what I got:

```
==================
WARNING: DATA RACE
Write at 0x00c00015c000 by goroutine 8:
  rely-evstore.(*AtomicCircularBuffer2).SaveEvent()
      /home/filippo/Documents/Vertex/rely-eventStore/atomiccircularbuffer2.go:47 +0xb4
  rely-evstore.TestRace()
      /home/filippo/Documents/Vertex/rely-eventStore/circularbuffer_test.go:26 +0x1c4
  testing.tRunner()
      /home/filippo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /home/filippo/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.24.1.linux-amd64/src/testing/testing.go:1851 +0x44

Previous read at 0x00c00015c000 by goroutine 9:
  rely-evstore.(*AtomicCircularBuffer2).eventMatchesFilter()
      /home/filippo/Documents/Vertex/rely-eventStore/atomiccircularbuffer2.go:121 +0x70f
  rely-evstore.(*AtomicCircularBuffer2).QueryEvents()
      /home/filippo/Documents/Vertex/rely-eventStore/atomiccircularbuffer2.go:97 +0x233
  rely-evstore.TestRace.func1()
      /home/filippo/Documents/Vertex/rely-eventStore/circularbuffer_test.go:21 +0x10f

....

==================
--- FAIL: TestRace (0.61s)
    testing.go:1490: race detected during execution of test
FAIL
exit status 1
FAIL    rely-evstore    0.620s
```

This confirmed an hypothesis I just had: even though you are using atomic to guard the access to the indices, you are not guarding the access to the slice, which is the most important.

When you do a write like

```golang
slice[i] = x
```

and at the same time a read

```golang
y = slice[j]
```

even though `i != j`, there is a data race.

--- a/atomiccircularbuffer2.go
+++ b/atomiccircularbuffer2.go
@@ -12,13 +12,12 @@ import (
 // AtomicCircularBuffer2 is an optimized, lock-free, fixed-size circular buffer for storing Nostr events.
 // It efficiently manages ephemeral events with a fixed memory footprint and automatic
 // event replacement when full using atomic operations for thread safety.
-// This implementation eliminates the explicit tail pointer and uses slice-based query results.
-// It uses atomic pointers to events to prevent data races between concurrent reads and writes.
+// This implementation uses atomic pointers to events to prevent data races between concurrent reads and writes.
 type AtomicCircularBuffer2 struct {
-	buffer []atomic.Pointer[nostr.Event]
-	head   uint64 // atomic - position to write next event
-	count  uint64 // atomic - number of events in buffer
-	size   uint64 // fixed size of the buffer
+	buffer []*atomic.Pointer[nostr.Event]
+	head   atomic.Uint64 // position to write next event
+	size   uint64        // fixed size of the buffer
+	count  atomic.Uint64 // number of events in buffer
 }
 
 // NewAtomicCircularBuffer2 creates a new AtomicCircularBuffer2 with the specified capacity.
@@ -27,7 +26,10 @@ func NewAtomicCircularBuffer2(capacity int) *AtomicCircularBuffer2 {
 		panic("capacity must be greater than 0")
 	}
 
-	buffer := make([]atomic.Pointer[nostr.Event], capacity)
+	buffer := make([]*atomic.Pointer[nostr.Event], capacity)
+	for i := range buffer {
+		buffer[i] = &atomic.Pointer[nostr.Event]{}
+	}
 
 	return &AtomicCircularBuffer2{
 		buffer: buffer,
@@ -43,22 +45,20 @@ func (cb *AtomicCircularBuffer2) SaveEvent(ctx context.Context, evt *nostr.Event
 		return errors.New("event cannot be nil")
 	}
 
-	// Create a copy of the event to store in the buffer
-	eventCopy := *evt
-
 	// Atomically get the current head
-	head := atomic.LoadUint64(&cb.head)
+	head := cb.head.Load()
 
-	// Store the event at the current head position using atomic operation
-	cb.buffer[head].Store(&eventCopy)
+	// Store the event directly at the current head position using atomic operation
+	// This avoids an unnecessary allocation by not creating a copy first
+	cb.buffer[head].Store(evt)
 
 	// Atomically increment the head with wrap-around
-	atomic.StoreUint64(&cb.head, (head+1)%cb.size)
+	cb.head.Store((head + 1) % cb.size)
 
 	// Update count atomically, capping at buffer size
-	count := atomic.AddUint64(&cb.count, 1)
+	count := cb.count.Add(1)
 	if count > cb.size {
-		atomic.CompareAndSwapUint64(&cb.count, count, cb.size)
+		cb.count.Store(cb.size) // Use Store directly instead of CompareAndSwap for better performance
 	}
 
 	return nil
@@ -69,11 +69,19 @@ func (cb *AtomicCircularBuffer2) SaveEvent(ctx context.Context, evt *nostr.Event
 // goroutine creation and channel operations.
 func (cb *AtomicCircularBuffer2) QueryEvents(ctx context.Context, filter nostr.Filter) ([]*nostr.Event, error) {
 	// Get a snapshot of the current state
-	count := atomic.LoadUint64(&cb.count)
+	count := cb.count.Load()
+	head := cb.head.Load()
 
 	// No events in buffer
 	if count == 0 {
 		return []*nostr.Event{}, nil
+	}
+
+	// Calculate tail position on demand
+	tail := uint64(0)
+	if count >= cb.size {
+		// Buffer is full, calculate tail from head
+		tail = (head + 1) % cb.size
 	}
 
 	// Apply limit from filter or use all events if no limit
@@ -82,16 +90,14 @@ func (cb *AtomicCircularBuffer2) QueryEvents(ctx context.Context, filter nostr.F
 		limit = filter.Limit
 	}
 
-	// Pre-allocate the result slice
+	// Pre-allocate the result slice with exact capacity
 	result := make([]*nostr.Event, 0, limit)
 
-	// Determine how many positions to scan - use min to get the smaller of count or size
-	scanCount := min(count, cb.size)
-
-	// Iterate through the buffer - no need for tail calculation
-	for i := uint64(0); i < scanCount; i++ {
+	// Start from the tail (oldest) and move towards head (newest)
+	for i := uint64(0); i < count; i++ {
+		idx := (tail + i) % cb.size
 		// Load the event atomically
-		evt := cb.buffer[i].Load()
+		evt := cb.buffer[idx].Load()
 		if evt != nil && cb.eventMatchesFilter(evt, filter) {
 			result = append(result, evt)
 			if len(result) >= limit {

--- a/atomiccircularbuffer2.go
+++ b/atomiccircularbuffer2.go
@@ -11,10 +11,11 @@ import (
 
 // AtomicCircularBuffer2 is an optimized, lock-free, fixed-size circular buffer for storing Nostr events.
 // It efficiently manages ephemeral events with a fixed memory footprint and automatic
-// oldest-event replacement when full using atomic operations for thread safety.
+// event replacement when full using atomic operations for thread safety.
 // This implementation eliminates the explicit tail pointer and uses slice-based query results.
+// It uses atomic pointers to events to prevent data races between concurrent reads and writes.
 type AtomicCircularBuffer2 struct {
-	buffer []nostr.Event
+	buffer []atomic.Pointer[nostr.Event]
 	head   uint64 // atomic - position to write next event
 	count  uint64 // atomic - number of events in buffer
 	size   uint64 // fixed size of the buffer
@@ -26,8 +27,10 @@ func NewAtomicCircularBuffer2(capacity int) *AtomicCircularBuffer2 {
 		panic("capacity must be greater than 0")
 	}
 
+	buffer := make([]atomic.Pointer[nostr.Event], capacity)
+
 	return &AtomicCircularBuffer2{
-		buffer: make([]nostr.Event, capacity),
+		buffer: buffer,
 		size:   uint64(capacity),
 		// head and count are initialized to 0 by default
 	}
@@ -40,15 +43,18 @@ func (cb *AtomicCircularBuffer2) SaveEvent(ctx context.Context, evt *nostr.Event
 		return errors.New("event cannot be nil")
 	}
 
+	// Create a copy of the event to store in the buffer
+	eventCopy := *evt
+
 	// Atomically get the current head
 	head := atomic.LoadUint64(&cb.head)
-	
-	// Store the event at the current head position
-	cb.buffer[head] = *evt
-	
+
+	// Store the event at the current head position using atomic operation
+	cb.buffer[head].Store(&eventCopy)
+
 	// Atomically increment the head with wrap-around
 	atomic.StoreUint64(&cb.head, (head+1)%cb.size)
-	
+
 	// Update count atomically, capping at buffer size
 	count := atomic.AddUint64(&cb.count, 1)
 	if count > cb.size {
@@ -63,9 +69,8 @@ func (cb *AtomicCircularBuffer2) SaveEvent(ctx context.Context, evt *nostr.Event
 // goroutine creation and channel operations.
 func (cb *AtomicCircularBuffer2) QueryEvents(ctx context.Context, filter nostr.Filter) ([]*nostr.Event, error) {
 	// Get a snapshot of the current state
-	head := atomic.LoadUint64(&cb.head)
 	count := atomic.LoadUint64(&cb.count)
-	
+
 	// No events in buffer
 	if count == 0 {
 		return []*nostr.Event{}, nil
@@ -80,21 +85,14 @@ func (cb *AtomicCircularBuffer2) QueryEvents(ctx context.Context, filter nostr.F
 	// Pre-allocate the result slice
 	result := make([]*nostr.Event, 0, limit)
 
-	// Calculate tail position (oldest event)
-	// When buffer is not full, tail is at position 0
-	// When buffer is full, tail is at position head (oldest event gets overwritten)
-	tail := uint64(0)
-	if count >= cb.size {
-		// Buffer is full, tail is at the same position as head
-		tail = head
-	}
+	// Determine how many positions to scan - use min to get the smaller of count or size
+	scanCount := min(count, cb.size)
 
-	// Iterate through the buffer starting from the oldest event
-	for i := uint64(0); i < count; i++ {
-		// Calculate position: start from tail and move forward
-		pos := (tail + i) % cb.size
-		evt := &cb.buffer[pos]
-		if cb.eventMatchesFilter(evt, filter) {
+	// Iterate through the buffer - no need for tail calculation
+	for i := uint64(0); i < scanCount; i++ {
+		// Load the event atomically
+		evt := cb.buffer[i].Load()
+		if evt != nil && cb.eventMatchesFilter(evt, filter) {
 			result = append(result, evt)
 			if len(result) >= limit {
 				break

--- a/bench-updated.md
+++ b/bench-updated.md
@@ -1,0 +1,47 @@
+# Benchmark Analysis Report: Atomic Operation Implementations
+
+## Summary
+
+Following a series of optimizations to the "Atomic2" implementation, and fixes for race conditions, this report details the comparative performance characteristics of three distinct code versions: "Original," "Atomic," and "Atomic2." The analysis covers single-write, concurrent-write, query, and mixed workload scenarios. The results demonstrate that the optimized Atomic2 implementation now delivers superior performance in nearly all scenarios, offering significant improvements in execution speed, memory efficiency, and scalability.
+
+## Benchmark Results Overview
+
+| Benchmark Name                | Implementation | ns/op   | B/op   | allocs/op |
+|-------------------------------|:--------------:|--------:|-------:|----------:|
+| BenchmarkSingleWrite          | Original       | 300.3   | 366    | 7         |
+|                               | Atomic         | 298.6   | 367    | 7         |
+|                               | Atomic2        | 297.9   | 363    | 7         |
+| BenchmarkConcurrentWrite      | Original       | 426.1   | 336    | 7         |
+|                               | Atomic         | 244.9   | 269    | 7         |
+|                               | Atomic2        | 193.5   | 260    | 7         |
+| BenchmarkQuery                | Original       | 28073   | 22534  | 4         |
+|                               | Atomic         | 28177   | 22252  | 4         |
+|                               | Atomic2        | 2028    | 1743   | 1         |
+| BenchmarkMixed                | Original       | 3890    | 4126   | 5         |
+|                               | Atomic         | 2597    | 3204   | 5         |
+|                               | Atomic2        | 374.9   | 370    | 4         |
+
+## Detailed Analysis
+
+### Single Write Operations
+
+All three implementations now perform comparably in single-write scenarios, with execution times hovering around 298–300 nanoseconds per operation. Memory usage and allocation counts are also nearly identical, indicating that the Atomic2 optimization has eliminated any previous overhead in this scenario.
+
+### Concurrent Write Operations
+
+The Atomic2 implementation stands out as the clear leader in concurrent-write performance, achieving an average execution time of 193.5 ns/op—approximately 21% faster than the Atomic implementation and over twice as fast as the Original. Additionally, Atomic2 demonstrates improved memory efficiency, utilizing only 260 bytes per operation.
+
+### Query Operations
+
+While all implementations exhibit similar performance for single and concurrent writes, the differences become pronounced in query-heavy workloads. Atomic2 completes queries in just 2028 ns/op, an order of magnitude faster than the Original (28073 ns/op) and Atomic (28177 ns/op) versions. Furthermore, Atomic2 allocates significantly less memory (1743 B/op vs. over 22000 B/op) and requires only a single allocation per operation, compared to four for the other implementations.
+
+### Mixed Workloads
+
+In mixed workload scenarios, which combine both read and write operations, Atomic2 maintains its substantial lead, completing operations in 374.9 ns/op—nearly seven times faster than the Atomic implementation and over ten times faster than the Original. Memory usage and allocation counts are also markedly lower with Atomic2.
+
+## Key Insights
+
+- **Performance Parity in Single-Write Scenarios:** The optimizations to Atomic2 have eliminated any previous performance penalty, bringing it on par with the Original and Atomic implementations.
+- **Superior Scalability:** Atomic2 demonstrates exceptional performance under concurrent operations, making it the optimal choice for high-concurrency environments.
+- **Query Efficiency:** The Atomic2 implementation remains significantly faster and more memory-efficient for query operations, delivering a substantial reduction in both execution time and resource consumption.
+- **Overall Efficiency:** Across all tested scenarios, Atomic2 consistently uses less memory and requires fewer allocations, reducing the workload on the garbage collector and improving overall system responsiveness.

--- a/bench.md
+++ b/bench.md
@@ -1,4 +1,4 @@
-## Benchmark Data Analysis Report
+# Benchmark Data Analysis Report
 
 This report analyzes the updated benchmark results from the `rely-evstore` Go package, executed on a 13th Gen Intel Core i7-13700H (Linux, amd64). The benchmarks compare three implementations (Original, Atomic, and Atomic2) across various operations: single writes, concurrent writes, queries, and mixed operations.
 

--- a/circularbuffer_test.go
+++ b/circularbuffer_test.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"fmt"
 	"sync"
-	"sync/atomic"
 	"testing"
 	"time"
 
@@ -324,8 +323,8 @@ func TestAtomicCircularBuffer2(t *testing.T) {
 
 	// Print buffer state before query
 	t.Logf("Buffer state: head=%d, count=%d, size=%d",
-		atomic.LoadUint64(&cb.head),
-		atomic.LoadUint64(&cb.count),
+		cb.head.Load(),
+		cb.count.Load(),
 		cb.size)
 
 	// Now buffer should have events 3-7 (5 events)

--- a/race_test.go
+++ b/race_test.go
@@ -1,0 +1,24 @@
+package main
+
+import (
+	"context"
+	"testing"
+
+	"github.com/nbd-wtf/go-nostr"
+)
+
+func TestRace(t *testing.T) {
+	ctx := context.Background()
+	ab := NewAtomicCircularBuffer2(1000)
+	event := createTestEvent("aaaaaaaaaaaaaaaaaaaaaaa", 1)
+
+	go func() {
+		for range 1000000 {
+			ab.QueryEvents(ctx, nostr.Filter{IDs: []string{"aaa"}})
+		}
+	}()
+
+	for range 1000000 {
+		ab.SaveEvent(ctx, event)
+	}
+}

